internal/getproviders/reattach/reattach.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package reattach
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/terraform/internal/addrs"
+)
+
+const TF_REATTACH_PROVIDERS = "TF_REATTACH_PROVIDERS"
+
+// ParseReattachProviders parses information used for reattaching to unmanaged providers out of a
+// JSON-encoded environment variable (TF_REATTACH_PROVIDERS).
+func ParseReattachProviders() (map[addrs.Provider]*plugin.ReattachConfig, error) {
+	in := os.Getenv(TF_REATTACH_PROVIDERS)
```

Comment:
Wouldn't it be easier to test things and reuse the function if we continued passing the value in as an argument, rather than read the environment variable?

Reply:
Yeah sure, we can avoid use of t.Setenv in tests that way. My rationalisation was that moving the call to os.Getenv into this package meant that the "TF_REATTACH_PROVIDERS" string wasn't floating around the codebase, but I can address that by making calling code use the new constant.

Reply:
I've refactored the function signature to require the new parameters in [12f8da6](https://github.com/hashicorp/terraform/pull/37634/commits/12f8da656670d3c103101d84f1af0937dd759667)